components/tikv_util/src/future.rs

```diff
@@ -72,25 +74,152 @@ where
 /// thread calling `wake()`.
 pub fn poll_future_notify<F: Future<Output = ()> + Send + 'static>(f: F) {
     let f: BoxFuture<'static, ()> = Box::pin(f);
-    let waker = Arc::new(BatchCommandsWaker(Mutex::new(Some(f))));
-    waker.wake();
+    let poller = Arc::new(PollAtWake {
+        f: UnsafeCell::new(Some(f)),
+        state: AtomicU8::new(IDLE),
+    });
+    PollAtWake::poll(&poller)
 }
 
-// BatchCommandsWaker is used to make business pool notifiy completion queues directly.
-struct BatchCommandsWaker(Mutex<Option<BoxFuture<'static, ()>>>);
+/// The future is not processed by any one.
+const IDLE: u8 = 0;
+/// The future is being polled by some thread.
+const POLLING: u8 = 1;
+/// The future is woken when being polled.
+const PENDING: u8 = 2;
```

Comment:
I was confused by the name `PENDING` at first. Personally, I think `NOTIFIED` is easier to understand...